src/main/java/com/example/web/controller/CheckoutController.java

```diff
@@ -108,7 +113,7 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
                     response.getWriter().write(outOfStockNames.toString());
                     return;
                 }
-                checkoutService.processCheckout(cart, userId,paymentMethodInt,recipientName, recipientPhone, deliveryAddress, shippingFeeDouble);
+                checkoutService.processCheckout(cart, userId,paymentMethodInt,recipientName, recipientPhone, deliveryAddress, shippingFeeDouble, appliedVoucherIds, shippingFeeFinal);
                 if (voucherIds != null) {
                     for (String vidStr : voucherIds) {
                         int voucherId = Integer.parseInt(vidStr);

```

Comment:
khai báo biến nên để ngoài vòng lặp vì biến sẽ được tạo lại mỗi lần lặp, làm tốn bộ nhớ và tài nguyên CPU

Reply:
oke tôi sẽ sửa lại, cảm ơn bạn góp ý 
```suggestion
 int voucherId;
for (String vidStr : voucherIds) {
    voucherId = Integer.parseInt(vidStr);
    userVoucherService.markAsUsed(voucherId, userId);
}
```